### PR TITLE
Big changes

### PR DESCRIPTION
**Pull request by aight8**
**Implement:**

**DraftEntity**
- Validate mutability value (enum)
- Remove static collection

**ContentBlock**
- setDepth - int cast and must be positive

**ContentState**
- getKeyBefore
- getKeyAfter
- getBlockBefore
- getBlockAfter
- implement entity map collection #6 
   - addEntity
   - createEntity
   - replaceEntityData
   - getEntity
   - getLastCreatedEntityKey
   - getEntityMap
   - setEntityMap
   - __setEntity
   - __removeEntity

*note: __ because it's not match draft.js ContentState class. But this should be removed. In draft.js you can have similar methods when you use the Immutable classes methods - but we don't have them here!*

**Encoding**
- convertToRaw (rename to convertFromDraftStateToRaw...)
- convertFromRaw
   - Checks the whole data structure and it's sub fields #4 for basic validness. Throws on very invalid states. Use default value otherwise.

**Validator**
Must be moved to another repository

**Tests**
Lot of tests for edge cases etc.

### CODE REVIEW

**Comment by robinvdvleuten** [comment]
o wow, thanks @aight8! What a nice beginning of a new year! The CI seems to be failing though, is that related to your code?

**Comment by robinvdvleuten** [comment]
Ah I see! You've added typehinting to the method's arguments but the library still supports php 5.6.

File: src/Model/Immutable/ContentState.php [hunk 1]
```diff
@@ -151,6 +162,80 @@ public function getLastBlock()
...
+    }
+
+    /**
+     * @param string $key
+     * @param string $relative
+     * @param bool $return_value
+     *
+     * @return ContentBlock|mixed|null|string
+     */
+    private function getRelativeBlock($key, $relative, $return_value = false)
```
Comment (robinvdvleuten) [inline src/Model/Immutable/ContentState.php]: Can you make this `$return_value`?

File: src/Model/Immutable/ContentState.php [hunk 2]
```diff
@@ -171,4 +256,63 @@ public function hasText()
...
+    }
+
+    /**
+     * @param string $key
+     *
+     * @return DraftEntity|null
+     */
+    public function getEntity($key)
+    {
+        return $this->entityMap[$key] ?? null;
```
Comment (robinvdvleuten) [inline src/Model/Immutable/ContentState.php]: Sorry, this must be the old way (`isset($this->entityMap[$key]) ? $this->entityMap[$key] : null;`) to work with 5.6

**Comment by aight8** [comment]
okey this PR will be a little bit bigger than excepted.
It's WIP, but feel free to comment so I can fix all those things.

What you think about the new Validator class?
It's strictly speaking more than a read-only validator since it can try to auto fix logical exceptions.

**Comment by robinvdvleuten** [comment]
@aight8 I see what you're doing with the validator class and config, but this is functionality not currently existing in the draft.js library. I try to mimic its functionality as close as possible and planning to add sub packages with any "extended" functionality. As I already have started a "renderer" package [here](https://github.com/webstronauts/draft-php-renderer), maybe it's an idea to move the validator to a draft-php-validator package for now. I really appreciate the effort you're putting in, but for now let's try to get the core functionality of Draft.js in this repo and add a list of any extending package to the readme for example.

I will try to get a list of all required / missing functionality as issues tomorrow in this repo to get a better overview of todos.

**Comment by aight8** [comment]
Ah I get it. You only want the draft.js functionality in this repository.
Yep than I must create an additional repo for this one.
The next what I try to implement is the Encodings export functionality to json again, so this one will a part of this repo one :)
I will create an additional repo for validation/filtering, since it's really needed to process valid draft.js structures in the backend. So it becomes very valuable for this project too :)

I must create a new PR since there are to much changes than.

BTW: How far you want to go with this library / where is the lowest common denominator between draft.js and this library? I don't see the exact border then. (Do you want create classes like "DraftEntityMutability" and "DraftDecorators" too?)
What is the vision of this project?

**Comment by robinvdvleuten** [comment]
@aight8 I agree with you that covering all functionality is not much of a value on the server side. The most required functionality is converting between the json state to the content state and vice versa. I've mimicked the class names as a check for myself to compare between js and php, it's ok for now to continue with "immutable" names and such. How to handle single function files in php must be looked at per case I guess. I will try to make a list of all missing but required functionalities tomorrow. Feel free to discuss or extend them then. 😄

**Comment by aight8** [comment]
Okey alright, that's good 👍 
convertToRaw method is implemented in the Encoding class and tested.
I must create a new PR sometime which excludes the Validator part.

File: spec/EncodingSpec.php [hunk 3]
```diff
@@ -11,18 +11,371 @@
...
+                            'offset' => 0,
+                            'length' => 1,
+                            'key' => 1,
+                        ],
+                    ],
+                ],
+            ],
+        ];
+
+        $raw = $this::convertToRaw($contentState);
```
Comment (robinvdvleuten) [inline spec/EncodingSpec.php]: Can you use -> instead of ::?

File: spec/EncodingSpec.php [hunk 4]
```diff
@@ -11,18 +11,371 @@
...
+
+        $theRaw = json_encode($raw->getWrappedObject());
+        $exceptedRaw = json_encode($exceptedRawData);
+
+        if ($theRaw !== $exceptedRaw) {
+            throw new \Exception('Comparison failed.');
+        }
```
Comment (robinvdvleuten) [inline spec/EncodingSpec.php]: The comparison should be like `$raw->getWrappedObject()->shouldBeLike($expectedRawData);`

File: spec/EncodingSpec.php [hunk 5]
```diff
@@ -11,18 +11,371 @@
...
+        $contentStateRaw = [
+            'entityMap' => [],
+            'blocks' => '', // <---
+        ];
+
+        $this::shouldThrow(InvalidRawException::class)
+            ->duringConvertFromRaw($contentStateRaw);
+    }
+
+    public function it_throws_exception_on_invalid_raw_2()
```
Comment (robinvdvleuten) [inline spec/EncodingSpec.php]: A better descriptive name would be better

File: spec/EncodingSpec.php (hunk 3 above)
Comment (robinvdvleuten) [inline spec/EncodingSpec.php]: Ah it's a static method

File: src/Encoding.php [hunk 6]
```diff
@@ -11,67 +11,252 @@
...
+            ];
+        }
+        $raw['entityMap'] = $entityMap;
+
+        $raw['blocks'] = array_map(function(ContentBlock $contentBlock) {
+            $inlineStyleRanges = [];
+            $entityRanges = [];
+
+            $allStyles = [];
+            foreach ($contentBlock->getCharacterList() as $characterMetadata) {
```
Comment (robinvdvleuten) [inline src/Encoding.php]: an array_reduce would be nicer than a loop.

File: src/Encoding.php [hunk 7]
```diff
@@ -11,67 +11,252 @@
...
+                    if ($hasStyle === false) continue;
+
+                    $styleRange = [
+                        'offset' => key($charList),
+                        'length' => null,
+                        'style' => $style,
+                    ];
+
+                    $styleLength = 0;
+                    do {
```
Comment (robinvdvleuten) [inline src/Encoding.php]: I am not very fond of these nested do/while loops

File: src/Encoding.php [hunk 8]
```diff
@@ -11,67 +11,252 @@
...
+     * @return CharacterMetadata[]
+     */
     public static function createCharacterList(array $inlineStyles, array $entities)
     {
         return array_map(function ($style, $index) use ($entities) {
             return new CharacterMetadata($style, $entities[$index]);
         }, $inlineStyles, array_keys($inlineStyles));
     }
 
+    private static function validateRange($ranges)
```
Comment (robinvdvleuten) [inline src/Encoding.php]: A better name would be assertRange

File: src/Encoding.php [hunk 9]
```diff
@@ -85,12 +270,25 @@ public static function decodeEntityRanges($text, array $ranges = null)
...
+     */
     public static function decodeInlineStyleRanges($text, array $ranges = null)
     {
         $styles = array_fill(0, strlen($text), []);
 
         if ($ranges) {
             foreach ($ranges as $range) {
+                self::validateRange($range);
+
+                if (!isset($range['style']) || !is_string($range['style']) || strlen($range['style']) < 1) {
```
Comment (robinvdvleuten) [inline src/Encoding.php]: Can this be moved to the validateRange method as well?

Comment (aight8) [inline src/Encoding.php]: validateRange is reused for decodeEntityRanges and this one have "key" instead of "style"

File: spec/EncodingSpec.php (hunk 4 above)
Comment (aight8) [inline spec/EncodingSpec.php]: This doesn't work either, I searched for a while how to assert two arrays but It looks like this doesn't exists :/
```
Call to a member function shouldBeLike() on array in ...
```

File: src/Encoding.php (hunk 6 above)
Comment (aight8) [inline src/Encoding.php]: You mean like this ?:
```
$allStyles = array_reduce(
                array_map(
                    function(CharacterMetadata $characterMetadata) {
                        return $characterMetadata->getStyle();
                    },
                    $contentBlock->getCharacterList()
                ),
                function($allStyles, array $styles) {
                    foreach ($styles as $style) {
                        $allStyles[$style] = $style;
                    }
                    return $allStyles;
                },
                []
            );
```

File: src/Encoding.php (hunk 7 above)
Comment (aight8) [inline src/Encoding.php]: It's now more verbose but without the inner loop. It's now readable.

File: src/Encoding.php (hunk 8 above)
Comment (aight8) [inline src/Encoding.php]: Yap renamed it

File: spec/EncodingSpec.php (hunk 5 above)
Comment (aight8) [inline spec/EncodingSpec.php]: I renamed all the test methods with the number at the end

File: spec/EncodingSpec.php (hunk 3 above)
Comment (aight8) [inline spec/EncodingSpec.php]: 👍

**Comment by aight8** [comment]
Okey done.
possible open points:
- $allStyles generation is now implemented with array_reduce, old variant is commented out. in my opinion it's now little bit too verbose but please check it out maybe you find a better way.
- Recommended test assert ```$raw->getWrappedObject()->shouldBeLike($expectedRawData);```  not working :( - maybe you know a better way, otherwise keep it for now since it does it job.

**Comment by aight8** [comment]
Multibyte strings are now handled correctly because it uses now mb_* php string functions.
So emoji's (4byte) and other 2 byte characters working now.
So there it creates now exactly as much character metadata how excepted.
But it worked until now since there was no modification operation (nor in toRaw neither in fromRaw), it just created too much character metadata objects in the character list.

File: src/Encoding.php [hunk 10]
```diff
@@ -11,69 +11,301 @@
...
+
+                    if ($hasStyle) {
+                        $lastCharacterHadCurrentStyle = true;
+                    } else {
+                        $lastCharacterHadCurrentStyle = false;
+                    }
+
+                } while (next($charList) !== false);
+
+                $inlineStyleRanges = array_merge($inlineStyleRanges, $currentStyleRanges);
```
Comment (robinvdvleuten) [inline src/Encoding.php]: Can you make this foreach an array_reduce?

File: src/Encoding.php [hunk 11]
```diff
@@ -11,69 +11,301 @@
...
 use Draft\Model\Immutable\ContentBlock;
+use Draft\Model\Immutable\ContentState;
 use Draft\Util\Keys;
 
 /**
  * @author Robin van der Vleuten <robin@webstronauts.co>
  */
 class Encoding
 {
+    public static function convertToRaw(ContentState $contentState)
```
Comment (robinvdvleuten) [inline src/Encoding.php]: This method has way too much complexity. Can you split it up to multiple smaller private static methods?

File: src/Encoding.php [hunk 12]
```diff
@@ -11,69 +11,301 @@
...
+        return $raw;
+    }
+
+    /**
+     * @param array $rawState
+     *
+     * @return ContentState
+     * @throws InvalidRawException
+     */
     public static function convertFromRaw(array $rawState)
```
Comment (robinvdvleuten) [inline src/Encoding.php]: Same counts for this method, must be split in smaller methods.

**Comment by robinvdvleuten** [comment]
@aight8 can you check out the StyleCI / TravisCI error output as well?

**Comment by robinvdvleuten** [comment]
@aight8 I think the PR is good enough for now 👍  Let's try to focus the PR to smaller chunks in the near future. There're still some small optimizations I would like to incorporate, but this will do for now, thanks 😄
